client/clientservice/src/state_snapshot_service.cpp

```diff
@@ -22,27 +27,405 @@ using vmware::concord::client::statesnapshot::v1::StreamSnapshotResponse;
+    }
+    response.response.template emplace<std::vector<std::unique_ptr<ResponseT>>>(std::move(responses));
+  }
+}
+
 Status StateSnapshotServiceImpl::GetRecentSnapshot(ServerContext* context,
                                                    const GetRecentSnapshotRequest* proto_request,
                                                    GetRecentSnapshotResponse* response) {
-  // TODO: Add implementation
-  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "GetRecentSnapshot");
+  std::chrono::milliseconds timeout = 5s;  // This is default timeout
+  if (context != nullptr) {
+    auto time_left =
+        gpr_time_to_millis(gpr_time_sub(gpr_now(context->raw_deadline().clock_type), context->raw_deadline()));
+    std::chrono::milliseconds curr_timeout{time_left};
+    if (curr_timeout.count() > 0) {
+      timeout = curr_timeout;
+    }
+  }
+
+  WriteConfig write_config{RequestConfig{false, 0, 1024 * 1024, timeout, "snapshotreq", "", false, true},
+                           bft::client::LinearizableQuorum{}};
+  StateSnapshotResponse snapshot_response;
+  grpc::Status return_status;
+  auto callback =
+      [&snapshot_response, &return_status, &write_config](concord::client::concordclient::SendResult&& send_result) {
+        ResponseType<StateSnapshotResponse> reponse;
+        getResponseSetStatus(std::move(send_result),
+                             reponse,
+                             return_status,
+                             write_config.request.correlation_id,
+                             false,
+                             "concord.client.clientservice.state_snapshot_service.GetRecentSnapshot.callback");
+        snapshot_response = std::get<StateSnapshotResponse>(reponse.response);
+      };
+
+  ReconfigurationRequest rreq;
+  StateSnapshotRequest cmd;
+  cmd.checkpoint_kv_count = 0;
+  cmd.participant_id = client_->getTRId();
+  rreq.command = cmd;
+  bft::client::Msg message;
+  concord::messages::serialize(message, rreq);
+  client_->send(write_config, std::move(message), callback);
+  if (return_status.ok() && snapshot_response.data.has_value()) {
+    response->set_snapshot_id(snapshot_response.data->snapshot_id);
+    response->set_event_group_id(snapshot_response.data->event_group_id);
+    response->set_key_value_count_estimate(snapshot_response.data->key_value_count_estimate);
+    auto* ledger_time = response->mutable_ledger_time();
+    *ledger_time = google::protobuf::util::TimeUtil::GetEpoch();
+  }
+  return return_status;
 }
 
 Status StateSnapshotServiceImpl::StreamSnapshot(ServerContext* context,
                                                 const StreamSnapshotRequest* proto_request,
                                                 ServerWriter<StreamSnapshotResponse>* stream) {
-  // TODO: Add implementation
-  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "StreamSnapshot");
+  Status status;
+
+  if ((proto_request != nullptr) && (proto_request->has_last_received_key())) {
```

Comment:
I was wondering about two things:

1. Can `proto_request` ever be nullptr? If yes, what does that mean? If it cannot be, maybe we can assert instead?
2. The client might not send `last_received_key`. If I am not missing something, we need to allow that.

Reply:
Thanks, @dartdart26, yes this needs to be changed, I will update this in the next commit.